File: julia/secret-handshake/_iterations/1/secret-handshake.jl
const actions = ["wink", "double blink", "close your eyes", "jump"]

function secret_handshake(code)
    handshake = [actions[i] for i in eachindex(actions) if code & (0b1 << (i - 1)) != 0]
    if code & 0b10000 != 0
        reverse!(handshake)
    end
    handshake
end

File: haskell/secret-handshake/_iterations/1/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits

actions :: [String]
actions = ["wink", "double blink", "close your eyes", "jump"]

handshake :: Int -> [String]
handshake n = reverseAction n . handshake' n actions $ []
  where
    handshake' :: Int -> [String] -> [String] -> [String]
    handshake' _ [] out = out
    handshake' n (a:as) out
             | n .&. 1 /= 0 = handshake' n' as (a:out)
             | otherwise = handshake' n' as out
                 where n' = shiftR n 1

    reverseAction :: Int -> [String] -> [String]
    reverseAction n out
                | n .&. 16 /= 0 = out
                | otherwise = reverse out
